docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,200 @@
+ * find candidate pod
+ * fail the pod
+ * wait until pod is terminated
+
+This means that there is no parallel killing of pods, and we will wait
+a single monitoring interval before killing an additional pod to allow
+for reclamation of the starved resource to take place.  By default, the
+`kubelet` will respect graceful termination period when doing eviction.
+
+The mechanism of finding a candidate pod for eviction is tied to an
+eviction strategy.  An eviction strategy is a compiled in plug-in that
+delegates finding a candidate pod.  The `kubelet` is configured to work
+against a particular strategy.  The default strategy if not specified is
+`greedy`.  Other strategies are anticipated in the future.
+
+```
+--eviction-strategy="greedy": Strategy to use when eviction thresholds have been met.
+```
+
+### Eviction Strategy: greedy
+
+The `greedy` strategy targets pods that are the largest consumers of the
+starved compute resource relative to their scheduling guarantee.
+
+It ranks pods within a quality of service tier in the following priority
+order:
+
+* `BestEffort` pods that consume the most of the starved resource are failed
+first
+* `Burstable` pods that consume the greatest amount of the starved resource
+relative to their request for that resource are killed first.  If no pod
+has exceeded its request, the strategy targets the largest consumer of the
+starved resource.
+* `Guaranteed` pods that consume the greatest amount of the starved resource
+relative to their request are killed first.  If no pod has exceeded its request,
+the strategy targets the largest consumer of the starved resource.
+
+### Alternative eviction strategies
+
+It is anticipated other eviction strategies may be used in the future.
+
+A number of potential decision planes to explore in the future:
+
+* target `BestEffort` pods that were most recently scheduled on the node
+* defer targeting pods that have data locality concerns, for example, its
+best to not evict a DB if it requires a local SSD
+* target pods with lower warm-up times before they are ready to accept traffic
+since they would be ready quicker on other nodes
+* etc.
+
+### Eviction best practices: DaemonSet
+
+It is never desired for a `kubelet` to evict a pod that was derived from
+a `DaemonSet` since the pod will immediately be recreated and rescheduled
+back to the same node.
+
+At the moment, the `kubelet` has no ability to distinguish a pod created
+from `DaemonSet` versus any other object.  If/when that information is
+available, the `kubelet` could pro-actively filter those pods from the 
+candidate set of pods provided to the eviction strategy plug-in.
+
+In general, it should be strongly recommended that `DaemonSet` not
+create `BestEffort` pods to avoid being identified as a candidate pod
+for eviction.
+
+## Improve node stability for BestEffort pods
+
+### Feasibility checks during kubelet admission
+
+The `kubelet` will reject `BestEffort` pods if any of its associated
```

Comment:
In case the scheduler messed up, the kubelet might have to reject `Guaranteed` pods as well.